lib/ecmarkdown.js

```diff
@@ -1,11 +1,357 @@
 'use strict';
-var parser = require('./generated-parser.js');
+
 var escapeHtml = require('escape-html');
+var beautify = require('./beautify.js');
 
-exports.list = function (ecmarkdown) {
-    return parser.parse(ecmarkdown.trim() + '\n', { startRule: 'list' });
+exports.document = function (ecmarkdown) {
+  return beautify(emitDocument(parse(ecmarkdown)));
 };
 
 exports.fragment = function (ecmarkdown) {
-    return parser.parse(escapeHtml(ecmarkdown.trim()), { startRule: 'fragment' });
+  return beautify(emitFragment(parse(escapeHtml(ecmarkdown))));
 };
+
+
+var lexemes = {
+  star: /\*/,
+  underscore: /_/,
+  tick: /`[^`\n]*`/,
+  pipe: /\|[^|\n]*\|/,
+  string: /"[^"\n]*"/,
+  tilde: /~/,
+  parabreak: /\n\n/,
+  linebreak: /\n/,
+  list: /[ \t]*\d+\. /,
+  whitespace: /[\s]+/,
+  comment: /<!--[\w\W]*-->/,
+  tag: /<[\/\w].*>/,
+  chars: /[^*_"~\s\n][^*_`"~<|\s\n]*/
+};
+
+var tokNames = Object.keys(lexemes);
+var tokRe = new RegExp(
+  tokNames.map(function (key) {
+    return '(^' + lexemes[key].toString().slice(1, -1) + ')';
+  })
+  .join('|')
+);
+
+var wrappers = {
+  star: 'emu-val',
+  underscore: 'var',
+  tick: 'code',
+  string: 'code',
+  tilde: 'emu-const'
+};
+
+function parse(emd) {
+  var t = new Tokenizer(emd);
+  var res = parseDocument();
+
+  return res;
+
+  function parseChars(inList, fmtStack) {
+    var node = { name: 'chars', contents: '' };
+    var tok = t.peek();
+
+    while (tok.name !== 'EOF' &&
+          (isChars(tok) || isWhitespace(tok) || isInvalidFormat(t.previous(), tok, t.peek(2), fmtStack))
+    ) {
+      t.next();
+      node.contents += tok.contents;
+      tok = t.peek();
+    }
+
+    return node;
+  }
+
+  function isInvalidFormat(prev, cur, next, fmtStack) {
+    if (!isFormatToken(cur)) {
+      return false;
+    }
+
+    return (fmtStack.indexOf(cur.name) === -1 && (isAlphaNumeric(prev) || !isAlphaNumeric(next))) ||
+           (fmtStack.indexOf(cur.name) > -1 && (!isAlphaNumeric(prev) || isAlphaNumeric(next)));
+  }
+
+  function isFormatToken(tok) {
+    return tok.name === 'star' || tok.name === 'underscore' || tok.name === 'tilde';
+  }
+
+  function isWhitespace(node) {
+    return node.name === 'whitespace' || node.name === 'linebreak';
+  }
+
+  function isAlphaNumeric(node) {
+    if (!node.contents) {
+      return false;
+    }
+    return !!node.contents.match(/^[\w\d]*$/);
+  }
+
+  function isChars(node) {
+    return node.name === 'chars';
+  }
+
+  function parseFormat(format, inList, fmtStack) {
```

Comment:
Yeah I will probably just add jsdoc once I am done with the refactoring (avoiding shared state will change just about every function signature).